File: backend/app/schemas/rams.py
"""RAMS schemas (PRD §8, §9)."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class RamsCreate(BaseModel):
    """Step 1-3 form payload. Validation mirrors the 4-step wizard."""

    # Step 1 — Project
    project_name: str = Field(min_length=3, max_length=300)
    site_address: str | None = Field(default=None, max_length=500)
    client_name: str | None = Field(default=None, max_length=300)
    project_reference: str | None = Field(default=None, max_length=100)
    start_date: str | None = Field(default=None, max_length=20)
    planned_duration: str | None = Field(default=None, max_length=100)

    # Step 2 — Work
    work_description: str = Field(min_length=10, max_length=5000)
    work_location: str | None = Field(default=None, max_length=500)
    materials: str | None = Field(default=None, max_length=2000)
    equipment: str | None = Field(default=None, max_length=2000)
    plant: str | None = Field(default=None, max_length=2000)
    tools: str | None = Field(default=None, max_length=2000)
    personnel: str | None = Field(default=None, max_length=2000)
    ppe: str | None = Field(default=None, max_length=2000)

    # Step 3 — Controls
    known_hazards: str | None = Field(default=None, max_length=3000)
    site_restrictions: str | None = Field(default=None, max_length=2000)
    existing_controls: str | None = Field(default=None, max_length=3000)
    emergency_info: str | None = Field(default=None, max_length=2000)
    additional_notes: str | None = Field(default=None, max_length=2000)


class RamsOut(BaseModel):
    id: str
    document_number: str | None = None
    project_name: str
    site_address: str | None = None
    client_name: str | None = None
    project_reference: str | None = None
    status: str
    input_data: dict[str, Any]
    generated_data: dict[str, Any] | None = None
    docx_path: str | None = None
    pdf_path: str | None = None
    generation_error: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class RamsListOut(BaseModel):
    items: list[RamsOut]
    total: int

File: frontend/lib/rams.ts
"use client";

import { apiFetch } from "@/lib/api";

export interface RamsInput {
  project_name: string;
  site_address: string;
  client_name: string;
  project_reference: string;
  start_date: string;
  planned_duration: string;
  work_description: string;
  work_location: string;
  materials: string;
  equipment: string;
  plant: string;
  tools: string;
  personnel: string;
  ppe: string;
  known_hazards: string;
  site_restrictions: string;
  existing_controls: string;
  emergency_info: string;
  additional_notes: string;
}

export interface Rams {
  id: string;
  document_number: string | null;
  project_name: string;
  site_address: string | null;
  client_name: string | null;
  project_reference: string | null;
  status: string;
  input_data: Record<string, unknown>;
  generated_data: Record<string, unknown> | null;
  docx_path: string | null;
  pdf_path: string | null;
  generation_error: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export const EMPTY_INPUT: RamsInput = {
  project_name: "",
  site_address: "",
  client_name: "",
  project_reference: "",
  start_date: "",
  planned_duration: "",
  work_description: "",
  work_location: "",
  materials: "",
  equipment: "",
  plant: "",
  tools: "",
  personnel: "",
  ppe: "",
  known_hazards: "",
  site_restrictions: "",
  existing_controls: "",
  emergency_info: "",
  additional_notes: "",
};

export function listRams(limit = 20, offset = 0) {
  return apiFetch<{ items: Rams[]; total: number }>(
    `/api/rams?limit=${limit}&offset=${offset}`
  );
}

export function createRams(input: RamsInput) {
  // Send only non-empty fields to keep payloads lean
  const payload: Record<string, string> = {};
  for (const [k, v] of Object.entries(input)) {
    if (v && v.trim()) payload[k] = v.trim();
  }
  return apiFetch<Rams>("/api/rams", { method: "POST", body: JSON.stringify(payload) });
}

export function getRams(id: string) {
  return apiFetch<Rams>(`/api/rams/${id}`);
}

export async function deleteRams(id: string) {
  const { createClient } = await import("@/lib/supabase/client");
  const supabase = createClient();
  const {
    data: { session },
  } = await supabase.auth.getSession();
  if (!session) throw new Error("Not authenticated");
  const res = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/api/rams/${id}`, {
    method: "DELETE",
    headers: { Authorization: `Bearer ${session.access_token}` },
  });
  if (!res.ok) throw new Error(`Delete failed (${res.status})`);
}
